core/src/main/java/io/grpc/internal/GzipInflatingBuffer.java

```diff
@@ -0,0 +1,514 @@
+    }
+    return inflaterOutputEnd;
+  }
+
+  private void writeInflatedBufToOutputBuffer(CompositeReadableBuffer bufferToWrite) {
+    bufferToWrite.addBuffer(ReadableBuffers.wrap(inflaterOutput, 0, inflaterOutputEnd));
+    inflaterOutputEnd = 0;
+    inflaterOutput = null;
+  }
+
+  private boolean inflate(int bytesToInflate) throws DataFormatException, ZipException {
+    checkState(inflater != null, "inflater is null");
+
+    try {
+      int inflaterTotalIn = inflater.getTotalIn();
+      int n = inflater.inflate(inflaterOutput, inflaterOutputEnd, bytesToInflate);
+      int bytesConsumedDelta = inflater.getTotalIn() - inflaterTotalIn;
+      bytesConsumed += bytesConsumedDelta;
+      inflaterInputStart += bytesConsumedDelta;
+      crc.update(inflaterOutput, inflaterOutputEnd, n);
+      inflaterOutputEnd += n;
+
+      if (inflater.finished()) {
+        // Save bytes written to check against the trailer ISIZE
+        expectedGzipTrailerIsize = (inflater.getBytesWritten() & 0xffffffffL);
+        if (gzipMetadataReader.readableBytes() <= GZIP_HEADER_MIN_SIZE + GZIP_TRAILER_SIZE) {
+          // We don't have enough bytes to begin inflating a concatenated gzip stream, drop context
+          inflater.end();
+          inflater = null;
+        }
+        state = State.TRAILER;
+        // Eagerly parse trailer, if possible, to detect CRC errors
+        return processTrailer();
+      } else if (inflater.needsInput()) {
+        state = State.INFLATER_NEEDS_INPUT;
+      }
+
+      return true;
+    } catch (DataFormatException e) {
+      // Wrap the exception so tests can check for a specific prefix
+      throw new DataFormatException("Inflater data format exception: " + e.getMessage());
+    }
+  }
+
+  private boolean initializeInflater() {
+    if (inflater == null) {
+      inflater = new Inflater(true);
+    } else {
+      inflater.reset();
+    }
+    crc.reset();
+    int bytesRemainingInInflaterInput = inflaterInputEnd - inflaterInputStart;
+    if (bytesRemainingInInflaterInput > 0) {
+      inflater.setInput(inflaterInput, inflaterInputStart, bytesRemainingInInflaterInput);
+      state = State.INFLATING;
+    } else {
+      state = State.INFLATER_NEEDS_INPUT;
+    }
+    return true;
+  }
+
+  private boolean fill() {
+    checkState(inflater != null, "inflater is null");
+    checkState(inflaterInputStart == inflaterInputEnd, "inflaterInput has unconsumed bytes");
+    int bytesToAdd = Math.min(gzippedData.readableBytes(), INFLATE_BUFFER_SIZE);
+    if (bytesToAdd > 0) {
+      inflaterInputStart = 0;
+      inflaterInputEnd = bytesToAdd;
+      gzippedData.readBytes(inflaterInput, inflaterInputStart, bytesToAdd);
+      inflater.setInput(inflaterInput, inflaterInputStart, bytesToAdd);
+      state = State.INFLATING;
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  private boolean processHeaderMagic() throws ZipException {
+    if (gzipMetadataReader.readableBytes() >= UNSIGNED_SHORT_SIZE) {
```

Comment:
It'd be nice if these checks were generally inverted. That would move the small case early (so it's easier to see what case it relates to) and it would allow avoiding the need for an else entirely (since the if will return).

Reply:
Done